### DISCUSSION TITLE
Coverting Data File to .tesr

### DISCUSSION THREAD

**Discussion by lkuna24** [neperfepx/neper#349] [Q&A] [answered]
Hey all,
I have a cellular automata code I am using for polycrystalline solidification simulations and would like to take my output and attempt to create a mesh from it.  The tutorial found here: https://neper.info/tutorials/tesr-mesh/#generating-a-mesh-msh seems to be well written and how I should approach the problem, but I can't seem to find the AlLi.dat (or anything like it).  If I knew the formatting of this data file, I could probably format my CA output in a similar manner and attempt to create a mesh.  My outputs do include orientation data as well, so at this point it is just a matter of formatting.  Is this file, or a sample representative file available for distribution?

**Comment by prsdshiva** [comment]
This [paper](https://www.sciencedirect.com/science/article/pii/S0045782517307028?via%3Dihub#d1e7868) mentions the Al data was obtained from [here.](https://www.sciencedirect.com/science/article/pii/S0921509309004511#ec-research-data) 

Does this lead you to a downloadable .dat file?

**Reply by rquey** [reply]
It is not the same data.

**Comment by rquey** [comment, marked as answer]
I will update the tutorial, but the `.dat` file is just a binary file (or it could be an ascii file) listing the grain ids of the voxels (x varying first). If the volume has X x Y x Z voxels, then there should be the same number of values in the file. The file format (binary or ascii) is to be specified in the tesr file as described in the tutorial.

The `.dat` file is also the raw data the Neper can produce using `-format raw -tesrformat ascii|binary16|...`. So, here is how you can get a sample `dat`/`raw`/`tesr` file:
```
neper -T -n 10 -for raw -tesrformat ascii
neper -T -n 10 -for raw -tesrformat binary16
```
(I've made a small fix in version 4.2.1-18).

**Reply by eugenianietov** [reply]
Hi! I have been trying to read some XRay tomography data post-processed with Dream3D following the aforementioned tutorial. I generated a .dat file with one row per voxel in my volume, indicating the grainID. My volume is 201 x 243 x 84 voxels and my step size is 5um, so my .dat file has 4102812 lines, with the grainID being either an integer or 0. My raster file is the following:

***tesr
 **format
    2.1
 **general
    3
    201 243 84
    5e-6 5e-6 5e-6
 **data
   binary16
  *file GrainId.dat
***end

When I try to load the .tesr file, I get a message when parsing the .dat file that states that my file contains more data than needed. Any idea of what can be wrong?

Thanks a lot in advance!!

Best,
Eugenia

**Reply by rquey** [reply]
It may have to do with how you wrote the `GrainId.dat` file.  You define it as `binary16` file, but you also say *"my .dat file has 4102812 lines"*.  A binary file should not have lines or line breaks. It should just be data.

**Reply by eugenianietov** [reply]
Okay that makes sense, my file is an ascii file, so I changed the .tesr file to "**data ascii", and now the error is directly "Aborted (core dumped)"...

**Reply by rquey** [reply]
Difficult to say without the `.dat` file. You can try to generate one with Neper and compare:
```
neper -T -n 3 -for raw -tesrsize 5:6:7 -o d
```
You can visualize like this:
```
neper -V d.tesr -print img
```
![image](https://github.com/neperfepx/neper/assets/2659376/d31605ad-8549-43e8-8de7-ec1e1bd0c08d)

**Reply by eugenianietov** [reply]
But that would create a .tesr file like the one we mentioned above, right? Not a .dat file, if I understood well. Would the .dat file result from adding '-stattesr cell' to the tessellation command? Thanks a lot!

**Reply by rquey** [reply]
`-for raw` will print a `.raw` (equivalent to your `.dat`) and a `.tesr` calling the `.raw`, so, exactly what you are doing.
```
$ neper -T -n 3 -for raw -tesrsize 5:6:7 -o d

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.6.1-3
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 8 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : Loading initialization file `/home/rquey/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] -tesrformat ascii
Info   : [com line] -n 3 -for raw -tesrsize 5:6:7 -o d
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellation...
Info   :   - Setting seeds... 
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   : Rasterizing tessellation...
Info   : Writing results...
Info   :     [o] Writing file `d.tesr'...
Info   :     [o] Writing file `d.raw'...
Info   :     [o] Wrote file `d.raw'.
Info   :     [o] Wrote file `d.tesr'.
Info   : Elapsed time: 0.004 secs.
========================================================================
```

**Reply by eugenianietov** [reply]
Okay, I finally figured it out. First, the .dat file (in ASCII format) should be a single line, with space-separated integers indicating the grainId of every voxel, with 0 for the empty ones. Also, I found that the **cell argument is mandatory in the .tesr file; otherwise, neper will abort. I provide my test files for a 50x50x50 voxels volume, in case someone finds them useful.

Thanks Mr. Quey, for taking the time!!

Eugenia
![test_505050](https://github.com/neperfepx/neper/assets/26228093/300650a5-a630-4b8a-b625-18f0bf3e53f0)
[test_505050.zip](https://github.com/neperfepx/neper/files/11705408/test_505050.zip)

**Reply by Anmol-Bakshi** [reply]
Can you please tell how to get ASCII format from EBSD image?
